README.md
# passport-google-verify-token

[![Build Status](https://dev.azure.com/abdouslayne/Personal%20Projects/_apis/build/status/passport-google-verify-token?branchName=master)](https://dev.azure.com/abdouslayne/Personal%20Projects/_build/latest?definitionId=2?branchName=master)
[![npm version](https://badge.fury.io/js/passport-google-verify-token.svg)](https://badge.fury.io/js/passport-google-verify-token)
![License](https://img.shields.io/npm/l/passport-facebook-token.svg)


[Passport](http://passportjs.org/) strategy for authenticating with [Google](http://www.google.com/)
access tokens using the OAuth 2.0 API.

This module lets you authenticate using Google in your Node.js applications.
By plugging into Passport, Google authentication can be easily and
unobtrusively integrated into any application or framework that supports
[Connect](http://www.senchalabs.org/connect/)-style middleware, including
[Express](http://expressjs.com/).

## Installation

    $ npm install passport-google-verify-token

## Usage

### Configure Strategy

The Google authentication strategy leverages the [Google Auth Library for Node.js](https://github.com/googleapis/google-auth-library-nodejs) to authenticates users. 
Applications must supply a `verify` callback which accepts the `idToken` or `access_token`
coming from the user to be authenticated, and then calls the `done` callback
supplying a `parsedToken` (with all its information in visible form) and the
`googleId`.

```js
var GoogleTokenStrategy = require("passport-google-verify-token").Strategy;

passport.use(new GoogleTokenStrategy({
      clientID: '12345.abcdefghijkl.apps.googleusercontent.com'// Specify the CLIENT_ID of the backend
     // If other clients (such as android / ios apps) also access the google api:
     // audience: [CLIENT_ID_FOR_THE_BACKEND, CLIENT_ID_ANDROID, CLIENT_ID_IOS, CLIENT_ID_SPA]
    },
    function(parsedToken, googleId, done) {
      User.findOrCreate(..., function (err, user) {
        done(err, user);
      });
    }
  ));
```
When verifying an idToken, the Google Auth library `verifyIdToken()` function is called, and the authentication is finished. When an `access_token` is passed, however, two steps have to be made:
1. The Google Auth `getTokenInfo()` function is called. This is to verify that the token is valid and not expired.
2. A request to `/oauth2/v3/userinfo` is sent. `access_token`s require this second step in order to get the same user information `idToken`s return. 

### Authenticate Requests

Use `passport.authenticate()`, specifying the `'google-verify-token'` strategy, to authenticate requests.

```js
app.post('/auth/google/token',
  passport.authenticate('google-verify-token'),
  function (req, res) {
    // do something with req.user
    res.send(req.user? 200 : 401);
  }
);
```

Or using Sails framework:

```javascript
// api/controllers/AuthController.js
module.exports = {
  facebook: function(req, res) {
    passport.authenticate('google-verify-token', function(error, user, info) {
      // do stuff with user
      res.ok();
    })(req, res);
  }
};
```

### Client Requests

Clients can send requests to routes that use google-verify-token authentication using query parms, body, or HTTP headers. Clients will need to transmit the `access_token` or the `id_token` that are received from Google after user logs in.

#### Sending access_token as a Query parameter

```
GET /auth/google/token?access_token=<TOKEN_HERE>
```

#### Sending access token as an HTTP header

Clients can choose to send the access token using the Oauth2 Bearer token (RFC 6750) compliant format

```
GET /resource HTTP/1.1
Host: server.example.com
Authorization: Bearer base64_access_token_string
```


#### Sending access token as an HTTP body

Clients can transmit the access token via the body

```
POST /resource HTTP/1.1
Host: server.example.com

access_token=base64_access_token_string
```
  

## Credits

  - [Abdou BOUROUBI](http://github.com/abouroubi)

## Thanks

This project is inspired by [Juanma Reyes](http://github.com/jmreyes)' [Passport Google Id Token](https://github.com/jmreyes/passport-google-id-token).

## License

The MIT License (MIT)

Copyright (c) 2018 Abdou Bouroubi

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog


## v3.0.0

Adds the ability to verify idTokens OR accessTokens.

### Breaking Changes
Update `google-auth-library` to version 8.7.0

Validate `acces_token` as parameter with `googleAuthClient.getTokenInfo` instead of `googleAuthClient.verifyIdToken`


## v2.0.0

Well hello 2.0 🎉 

### Breaking Changes

#### Update Google Auth Library
Update `google-auth-library` to version 7.0.4


## v1.x
Initial release

tests/strategy.test.ts
import * as chai from 'chai';
import { Strategy } from '../src';

// tslint:disable-next-line:no-var-requires
const passport = require('chai-passport-strategy');

const chaiPassport = chai.use(passport);
const expect = chai.expect;

declare global {
  namespace Chai {
    // tslint:disable-next-line:interface-name
    interface ChaiStatic {
      passport: any;
    }
  }
}

describe('Strategy', () => {
  const verify = (parsedToken: any, googleId: any, done: (...args: any[]) => void) => {
    return done(null, { id: '1234' }, { scope: 'read' });
  };

  const mockVerifyGoogleTokenSuccess = (idToken: string, clientID: string | [], cb: (...args: any[]) => void) => {
    const payload = { sub: 1 };
    cb(null, payload);
  };

  const mockVerifyGoogleTokenNoParsedToken = (idToken: string, clientID: string | [], cb: (...args: any[]) => void) => {
    cb(null, false, { message: 'Error message' });
  };

  const mockVerifyGoogleTokenError = (idToken: string, clientID: string | [], cb: (...args: any[]) => void) => {
    cb({ message: 'Error message' });
  };

  const mockToken = '123456790-POIHANPRI-KNJYHHKIIH';

  const strategy = new Strategy(
    {
      clientID: 'DUMMY_CLIENT_ID',
    },
    verify,
  );

  strategy.verifyGoogleToken = mockVerifyGoogleTokenSuccess;

  const strategyWClientIDArray = new Strategy(
    {
      clientID: ['DUMMY_CLIENT_ID_1', 'DUMMY_CLIENT_ID_2', 'DUMMY_CLIENT_ID', 'DUMMY_CLIENT_ID_3'],
    },
    verify,
  );

  strategyWClientIDArray.verifyGoogleToken = mockVerifyGoogleTokenSuccess;

  const strategyNoParsedToken = new Strategy(
    {
      clientID: 'DUMMY_CLIENT_ID',
    },
    verify,
  );

  strategyNoParsedToken.verifyGoogleToken = mockVerifyGoogleTokenNoParsedToken;

  const strategyTokenError = new Strategy(
    {
      clientID: 'DUMMY_CLIENT_ID',
    },
    verify,
  );

  strategyTokenError.verifyGoogleToken = mockVerifyGoogleTokenError;

  it('should be named google-verify-token', () => {
    expect(strategy.name).to.equal('google-verify-token');
  });

  it('should throw if constructed without a verify callback', () => {
    expect(() => {
      const s = new Strategy({});
    }).to.throw('GoogleVerifyTokenStrategy requires a verify function');
  });

  function performValidTokenTest(strategyObject: any, reqFunction: any) {
    let user: any;
    let info: any;

    before(done => {
      chaiPassport.passport
        .use(strategyObject)
        .success((u: any, i: any) => {
          user = u;
          info = i;
          done();
        })
        .request(reqFunction)
        .authenticate();
    });

    it('should supply user', () => {
      expect(user).to.be.an('object');
      expect(user.id).to.equal('1234');
    });

    it('should supply info', () => {
      expect(info).to.be.an('object');
      expect(info.scope).to.equal('read');
    });
  }

  describe('handling a request with id_token as query parameter', () => {
    performValidTokenTest(strategy, (req: any) => {
      req.query = { id_token: mockToken };
    });
  });

  describe('handling a request with access_token as query parameter', () => {
    performValidTokenTest(strategy, (req: any) => {
      req.query = { access_token: mockToken };
    });
  });

  describe('handling a request with id_token as body parameter', () => {
    performValidTokenTest(strategy, (req: any) => {
      req.body = { id_token: mockToken };
    });
  });

  describe('handling a request with access_token as body parameter', () => {
    performValidTokenTest(strategy, (req: any) => {
      req.body = { access_token: mockToken };
    });
  });

  describe('handling a request with bearer token in authorization header', () => {
    performValidTokenTest(strategy, (req: any) => {
      req.headers = { authorization: 'Bearer ' + mockToken };
    });
  });

  describe('handling a valid request with clientID array in strategy options', () => {
    performValidTokenTest(strategyWClientIDArray, (req: any) => {
      req.body = { access_token: mockToken };
    });
  });

  // Failing tests
  function performFailTokenTest(strategyObject: any, reqFunction: any) {
    let error: any;

    before(done => {
      chaiPassport.passport
        .use(strategyObject)
        .fail((u: any) => {
          error = u;
          done();
        })
        .request(reqFunction)
        .authenticate();
    });

    it('should error object exists', () => {
      expect(error).to.be.an('object');
    });
  }

  describe('handling a request with no id token', () => {
    performFailTokenTest(strategy, (req: any) => {
      return;
    });
  });

  describe('handling a valid request with error', () => {
    performFailTokenTest(strategyTokenError, (req: any) => {
      req.body = { access_token: mockToken };
    });
  });

  describe('handling a valid request with no parsed token', () => {
    performFailTokenTest(strategyNoParsedToken, (req: any) => {
      req.body = { access_token: mockToken };
    });
  });
});


src/strategy.ts
/**
 * Module dependencies.
 */
import { OAuth2Client } from 'google-auth-library';
import { Strategy } from 'passport-strategy';
import got from 'got'

/**
 * `Strategy` constructor.
 *
 * The Google authentication strategy authenticates requests by verifying the
 * signature and fields of the token.
 *
 * Applications must supply a `verify` callback which accepts the `idToken`
 * coming from the user to be authenticated, and then calls the `done` callback
 * supplying a `parsedToken` (with all its information in visible form) and the
 * `googleId`.
 *
 * Options:
 * - `clientID` your Google application's client id (or several as Array)
 *
 * Examples:
 *
 * passport.use(new GoogleTokenStrategy({
 *     clientID: '12345.abcdefghijkl.apps.googleusercontent.com'// Specify the CLIENT_ID of the app that accesses the backend
 *    // Or, if multiple clients access the backend:
 *    //[CLIENT_ID_1, CLIENT_ID_2, CLIENT_ID_3]
 *   },
 *   function(parsedToken, googleId, done) {
 *     User.findOrCreate(..., function (err, user) {
 *       done(err, user);
 *     });
 *   }
 * ));
 *
 * @param {Object} options
 * @param {Function} verify
 * @api public
 */
export class GoogleTokenStrategy extends Strategy {
  public name: string;
  public googleAuthClient: OAuth2Client;
  public verify: (...args: any[]) => void;
  public clientID: string;
  public passReqToCallback: boolean;
  public audience: string[];

  constructor(options: (() => void) | any, verify?: (...args: any[]) => void) {
    super();

    if (typeof options === 'function') {
      verify = options;
      options = {};
    }

    if (!verify) {
      throw new Error('GoogleVerifyTokenStrategy requires a verify function');
    }

    this.passReqToCallback = options.passReqToCallback;

    this.clientID = options.clientID;

    this.name = 'google-verify-token';
    this.googleAuthClient = new OAuth2Client(this.clientID);
    this.verify = verify;
    this.audience = options.audience ? options.audience : options.clientID;
  }

  /**
   * Internal function that handles successes/failures
   * 
   * @param {any} err 
   * @param {any} parsedToken 
   * @param {any?} info
   */
  private done(err: any, parsedToken: any, info?: any) {
    if (err) {
      return this.fail({ message: err.message }, 401);
    }

    if (!parsedToken) {
      return this.fail(info);
    }

    const verified = (error: any, user: any, infoOnUser: any) => {
      if (error) {
        return this.error(error);
      }
      if (!user) {
        return this.fail(infoOnUser);
      }
      this.success(user, infoOnUser);
    };

    if (parsedToken.sub) this.verify(parsedToken, parsedToken.sub, verified);
    else {
      this.verify(parsedToken, parsedToken, verified)
    }
  }

  /**
   * Authenticate request by verifying the token
   *
   * @param {Object} req
   * @param {Object} options
   * @api protected
   */
  public authenticate(req: any, options: any) {
    options = options || {};

    const accessToken = this.paramFromRequest(req, 'access_token');
    const idToken =
      this.paramFromRequest(req, 'id_token') ||
      this.getBearerToken(req.headers);

    if (idToken) this.verifyGoogleIdToken(idToken)
    else if (accessToken) this.verifyGoogleAccessToken(accessToken)
    else {
      return this.fail({ message: 'no Google authentication token provided' }, 401);
    }
  }

  /**
   * Verify signature and token fields for an id_token
   *
   * @param {String} idToken
   * @param {String} clientID
   * @api protected
   */
  public verifyGoogleIdToken(idToken: string) {
    this.googleAuthClient.verifyIdToken(
      {
        audience: this.audience,
        idToken,
      },
      (err, loginTicket) => {
        if (err) {
          this.done(null, false, { message: err.message });
        } else if (loginTicket) {
          const payload = loginTicket.getPayload();
          this.done(null, payload);
        } else {
          this.done(null, false, { message: 'No login ticket retuned' });
        }
      },
    );
  }

  /**
   * Ensure getting token info for access token is successful.
   * 
   * @param {String} accessToken
   * @api protected
   */
  public verifyGoogleAccessToken(accessToken: string) {
    this.googleAuthClient.getTokenInfo(accessToken).then((tokenInfo) => {
      if (!tokenInfo) {
        this.done(null, false, {
          message: 'invalid access token'
        })

        return
      }

      if (tokenInfo.expiry_date < Date.now()) {
        this.done(null, false, {
          message: 'access token expired'
        })

        return
      }

      // Now we have to get the userinfo from the token
      got.get(`https://www.googleapis.com/oauth2/v3/userinfo?access_token=${accessToken}`).then(userinfo => {
        this.done(null, userinfo)
      }).catch(e => {
        this.done(null, false, {
          message: 'failed to get userinfo'
        })
      })
    }).catch((e) => {
      this.done(null, false, {
        message: e.message
      })
    })
  }

  /**
   * Gets the id token value from req using name for lookup in req.body, req.query,
   * and req.params.
   * @param {express.Request} req
   * @param {string} name  the key to use to lookup id token in req.
   * @api protected
   */
  private paramFromRequest(req: any, name: string) {
    const body = req.body || {};
    const query = req.query || {};
    const params = req.params || {};
    const headers = req.headers || {};
    if (body[name]) {
      return body[name];
    }
    if (query[name]) {
      return query[name];
    }
    if (headers[name]) {
      return headers[name];
    }
    return params[name] || '';
  }

  private getBearerToken(headers: any) {
    if (headers && headers.authorization) {
      const parts = headers.authorization.split(' ');
      return parts.length === 2 && parts[0] === 'Bearer' ? parts[1] : undefined;
    }
  }
}
